File: hydra-scala/src/main/scala/hydra/lib/pairs.scala
package hydra.lib

object pairs:
  /** Map over both elements of a pair. */
  def bimap[A, B, C, D](f: A => C)(g: B => D)(p: (A, B)): (C, D) = (f(p._1), g(p._2))

  /** Get the first element of a pair. */
  def first[A, B](p: (A, B)): A = p._1

  /** Get the second element of a pair. */
  def second[A, B](p: (A, B)): B = p._2

File: hydra-lisp/hydra-clojure/src/main/clojure/hydra/lib/pairs.clj
(ns hydra.lib.pairs)

;; Pairs are 2-element lists: (list a b)

;; bimap :: (a -> c) -> (b -> d) -> Pair a b -> Pair c d
(def hydra_lib_pairs_bimap
  "Map over both elements of a pair."
  (fn [f] (fn [g] (fn [p] (list (f (first p)) (g (second p)))))))

;; first :: Pair a b -> a
(def hydra_lib_pairs_first
  "Get the first element of a pair."
  (fn [p] (first p)))

;; second :: Pair a b -> b
(def hydra_lib_pairs_second
  "Get the second element of a pair."
  (fn [p] (second p)))
